lib/random.gd

```diff
@@ -69,9 +69,84 @@ DeclareCategory( "IsRandomSource", IsComponentObjectRep );
 ##  </ManSection>
 ##  <#/GAPDoc>
 ##
-DeclareOperation( "Random", [IsRandomSource, IsList] );
+DeclareOperation( "Random", [IsRandomSource, IsListOrCollection] );
 DeclareOperation( "Random", [IsRandomSource, IsInt, IsInt] );
 
+##############################################################################
+##
+##  <#GAPDoc Label="InstallMethodWithGlobalRandomGen">
+##  <ManSection>
+##  <Func Name="InstallMethodWithGlobalRandomGen"
+##   Arg="opr,info[,famp],args-filts[,val],method"/>
+##
+##  <Description>
+##  A helper function which works the same as <Ref Func="InstallMethod"/>,
+##  The purpose of this method is to make is easier and less error-prone to
+##  install pairs of methods for
+##  <Ref Oper="Random"  Label="for random source and list"/> and
+##  <Ref Oper="PseudoRandom"/> which accept an explicit random source as
+##  first argument, or use the global random source.
+##
+##  The arguments to this function are identical to
+##  <Ref Func="InstallMethod"/> except that the first argument to
+##  <A>args-filts</A> must be IsRandomSource, and the <A>info</A> string
+##  is compulsory, and must begin 'for a random source and'.
+##  </Description>
+##  </ManSection>
+##  <#/GAPDoc>
+##
+BIND_GLOBAL("InstallMethodWithGlobalRandomGen", function(args...)
+    local str, filterpos, filtercopy, argscopy, i, func, info;
+
+    # Check we understand arguments
+    # Second value must be an info string
+    if not IsString(args[2]) then
+        ErrorNoReturn("Second argument must be an info string");
+    fi;
+
+    # Info strings always tend to begin 'for ', and here we want
+    # to be able to edit it, so we check.
+    if args[2]{[1..23]} <> "for a random source and" then
```

Comment:
I guess `StartsWith` is not available here?

Hmm... looking at `lib/read*.g`, it seems we used to read it later, but it was moved to `read1.g`. Testing this, the only reason for that seems to be that `matobj2.gd` references random sources... So we can fix this (and you can start using things like `BindGlobal` and `StartsWith` here) quite easily. This patch does it for me:

```diff
diff --git a/lib/read1.g b/lib/read1.g
index 185f74763..3843fdbfd 100644
--- a/lib/read1.g
+++ b/lib/read1.g
@@ -67,8 +67,6 @@ ReadLib( "info.gi"     );
 ReadLib( "assert.gi"   );
 ReadLib( "global.gi"   );

-ReadLib( "random.gd"   );
-
 ReadLib( "options.gd"  );
 ReadLib( "options.gi"  );

diff --git a/lib/read3.g b/lib/read3.g
index 51a34f289..40f2730c3 100644
--- a/lib/read3.g
+++ b/lib/read3.g
@@ -82,9 +82,6 @@ ReadLib( "unknown.gd"  );
 ReadLib( "word.gd"     );
 ReadLib( "wordass.gd"  );

-ReadLib( "matobj2.gd"  );
-ReadLib( "matobjplist.gd" );
-
 # files dealing with rewriting systems
 ReadLib( "rws.gd"      );
 ReadLib( "rwspcclt.gd" );
@@ -174,9 +171,13 @@ ReadLib( "integer.gi"  ); # needed for CoefficientsQadic
 ReadLib( "list.gi"     ); # was too early
 ReadLib( "set.gi"      );
 ReadLib( "wpobj.gi"    );
-##  ReadLib( "random.gd"   );
-##  ReadLib( "random.gi"   );
-##  ReadLib( "random.g"    );
+
+# random sources
+ReadLib( "random.gd"   );
+ReadLib( "random.gi"   );
+
+ReadLib( "matobj2.gd"  );
+ReadLib( "matobjplist.gd" );

 # files dealing with nice monomorphism
 # grpnice uses some family predicates, so fampred.g must be known
@@ -249,6 +250,3 @@ ReadLib("gasman.gd");

 # files dealing with function calling
 ReadLib("function.gd");
-
-# random sources
-ReadLib("random.gi");
```